tower/src/util/either.rs

```diff
@@ -17,58 +16,73 @@ use tower_service::Service;
 /// Both services must be of the same request, response, and error types.
 /// [`Either`] is useful for handling conditional branching in service middleware
 /// to different inner service types.
-#[pin_project(project = EitherProj)]
-#[derive(Clone, Debug)]
+#[derive(Clone, Copy, Debug)]
 pub enum Either<A, B> {
-    /// One type of backing [`Service`].
-    A(#[pin] A),
-    /// The other type of backing [`Service`].
-    B(#[pin] B),
+    #[allow(missing_docs)]
+    Left(A),
+    #[allow(missing_docs)]
+    Right(B),
 }
 
 impl<A, B, Request> Service<Request> for Either<A, B>
 where
     A: Service<Request>,
-    A::Error: Into<crate::BoxError>,
-    B: Service<Request, Response = A::Response>,
-    B::Error: Into<crate::BoxError>,
+    B: Service<Request, Response = A::Response, Error = A::Error>,
 {
     type Response = A::Response;
-    type Error = crate::BoxError;
-    type Future = Either<A::Future, B::Future>;
+    type Error = A::Error;
+    type Future = EitherResponseFuture<A::Future, B::Future>;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        use self::Either::*;
-
         match self {
-            A(service) => Poll::Ready(Ok(ready!(service.poll_ready(cx)).map_err(Into::into)?)),
-            B(service) => Poll::Ready(Ok(ready!(service.poll_ready(cx)).map_err(Into::into)?)),
+            Either::Left(service) => service.poll_ready(cx),
+            Either::Right(service) => service.poll_ready(cx),
         }
     }
 
     fn call(&mut self, request: Request) -> Self::Future {
-        use self::Either::*;
-
         match self {
-            A(service) => A(service.call(request)),
-            B(service) => B(service.call(request)),
+            Either::Left(service) => EitherResponseFuture {
+                kind: Kind::Left {
+                    inner: service.call(request),
+                },
+            },
+            Either::Right(service) => EitherResponseFuture {
+                kind: Kind::Right {
+                    inner: service.call(request),
+                },
+            },
         }
     }
 }
 
-impl<A, B, T, AE, BE> Future for Either<A, B>
+pin_project! {
+    /// Response future for [`Either`].
+    pub struct EitherResponseFuture<A, B> {
+        #[pin]
+        kind: Kind<A, B>
+    }
+}
+
+pin_project! {
+    #[project = KindProj]
+    enum Kind<A, B> {
+        Left { #[pin] inner: A },
+        Right { #[pin] inner: B },
+    }
+}
```

Comment:
nit, not a big deal: can these be combined into a single `pin_project!` block?
```suggestion
pin_project! {
    /// Response future for [`Either`].
    pub struct EitherResponseFuture<A, B> {
        #[pin]
        kind: Kind<A, B>
    }

    #[project = KindProj]
    enum Kind<A, B> {
        Left { #[pin] inner: A },
        Right { #[pin] inner: B },
    }
}
```

Reply:
Oh cool I didn't realize you could do this!

Reply:
i think i was actually wrong and you can't do this, this seems to have broken CI. my bad, sorry --- we should back this change out.